DataDog/dd-trace-dotnet
# Various CI improvements

## Summary of changes

- Avoids building samples that are unused
- Make running "test all package versions" in CI for specific sample v. quick
- Add some resilience to NuGet

## Reason for change

On PRs and master, we only test a limited subset of package versions due to speed and size considerations. We have a variable to allow testing all versions, but this always fails in CI, as we either run out of drive space trying to build hundreds (prob. thousands!) of samples, or we timeout. 

With the changes in this PR, you can tell CI to only build a single sample, and only run a single test. This takes integration test run times down to ~5 mins (instead of the usual 40 when running all tests, or 1hr+ if trying to do multi-package version)

## Implementation details

Update the Nuke targets, MSBuild _.proj_ file, docker-compose files, and yaml pipeline to allow taking `Filter` and `SampleName` into account. These can be specified when doing a "custom" `Run Pipeline` in Azure Devops.

Additional quality of life improvements:
- Removed case sensitivity for SampleName where possible (I couldn't work out how to do this properly in the msbuild _.proj_ file, so it's still best to use the correct casing, as the multi-package version stage requires it
- Don't build samples that require Docker if we're not running the Docker tests (to speed up CI)
- Don't build the "trimming test" app if not going to be tested
- Add `NUGET_EXPERIMENTAL_CHAIN_BUILD_RETRY_POLICY` (azdo) and `NUGET_CERT_REVOCATION_MODE=offline` (gitlab) to handle issues with cert chaining. See [here](https://learn.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu3028) for details

## Test coverage

Ran various tests, and works well! The following integration test run times are for a run where I ran comprehensive testing, but limited to only a single sample and a single test class:

![image](https://github.com/DataDog/dd-trace-dotnet/assets/18755388/2a83a5e5-32fc-4446-8da6-345bf452d246)

## Other details

After these changes, we should probably make it a general policy that any PR which "significantly" changes an integration, should do a manual run with the following settings:

- `perform_comprehensive_testing`: `true` - to test against all package versions
- `run_all_test_frameworks`: `true` - to test against all supported framework versions
- `TEST_FILTER`: `<SomeTestClass>` - the test class to run
- `TEST_SAMPLE_NAME`: `<Samples.MySample>` - the sample to build

You can set these variables [by going to Azure DevOps](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build?definitionId=54):
- Click **Run Pipeline**
- Under **branch/tag** set the name of your branch. Note that you _can't_ use the `pr/xyz` pseudo-branch here, it has to be your _real_ branch
- Under **Variables** set the values above, then click the **Back arrow**
- Under **Stages**, _optionally_ limit to only the stages you need. For example, to run Linux x64 integration tests, select the following stages, then click **Use selected stages**
  - `merge_commit_id`
  - `generate_variables`
  - `build_linux_tracer`
  - `build_linux_profiler`
  - `package_linux`
  - `integration_tests_linux`
- Click **Run**
